### PR TITLE
Update breakpad and apply the stackwalker improvements for bug 1682806 and bug 1683232

### PR DESCRIPTION
**Pull request by gabrielesvelto**
This applies a couple of out-of-tree patches that were developed for mozilla-central. The first improves breakpad's stackwalker to provide better results for AArch64 when encountering buggy CFI information. The second supports new AMD64 and X86 variable-sized contexts that were introduced when using recent CPUs on Windows 10 2020 H1&H2 versions; without this change the stack walker can't extract any frames from the stack.

### CODE REVIEW

**Comment by gabrielesvelto** [comment]
I tried adding profile data for the updated version but it doesn't work, probably because I'm using a different version of GCC. Will, how should we handle this? One thing I thought about is that I could create a set of minidumps on my machine that I'd know are free of sensitive information and then add it to the sources so we can also do the profile generation in CI. I'd cover all our use-cases Linux/macOS/Windows/Android and x86/x86-64/arm/arm64. Naturally I'd also have to check-in the symbol files so it would amount to a fair bit of data, but I think it's feasible. WDYT?

**Comment by willkg** [comment]
You have to build the profile _inside_ the container. I have some clunky instructions here:

https://github.com/mozilla-services/minidump-stackwalk#pgo-profile

I think this would take me a few hours to do, but I'm not going to get to it this year.

I think having minidumps in this repository that we could build profiles with could be cool. It might help for testing, too. I think we'd have to figure out symbols files. minidumps would need to be tied to modules for which we have symbols and something to serve them. Tecken expires symbols after 2 years, so unless we want to rewrite everything in two years we'd need symbols and something to serve them to this repository, too.

**Comment by willkg** [comment]
I haven't forgotten about this--it's in my todo list for this week.

**Comment by gabrielesvelto** [comment]
> I haven't forgotten about this--it's in my todo list for this week.

Don't worry for now, I'm trying to integrate a bunch of minidumps and symbol files to let the profile-generation happen during the build. Ignore this PR until I'm done :-)

**Comment by willkg** [comment]
You rock! Thank you so much for doing this!

Two groups of questions:

1. Is a macos-x86-64 and android-arm dumps enough? Should we collect other dumps to profile? If so, I'll add an issue for whatever other dumps we should have.
2. Is it possible for these dumps to go "stale"? Do we want to keep track of when they should be updated? If so, how do we update them?

**Comment by gabrielesvelto** [comment]
I'm collecting other dumps too, I want to cover all platforms. The formats didn't really change in years so I think we shouldn't have to worry about it.

**Comment by gabrielesvelto** [comment]
I squashed all the commits into one and finalized the training set. The resulting build should use PGO by default. This is going to be useful even when we rewrite this stuff as we won't have to rebuild a new PGO training set then. PGO only leads to small but consistent speedups as we had noticed before. This is ready for review!

File: README.rst [hunk 1]
```diff
@@ -103,15 +103,5 @@ Things to keep in mind:
...
-1. swap the ``PGOFLAGS`` lines to ``-fprofile-generate`` in
-   ``minidump-stackwalk/Makefile`` and ``bin/build-breakpad.sh``
-2. run ``make clean`` and ``make build``
-3. process 100 crash reports
-4. the profile should be in ``/app/pgo_profile/`` in the container and ``./pgo_profile/``
-   outside the container
-5. swap the ``PGOFLAGS`` lines back to ``-fprofile-use``
-6. commit profile to git
+By default both minidump-stackwalk and breakpad are built with PGO and an
+appropriate training set is included in the sources.
```
Comment (willkg) [inline README.rst:107]: Thank you so much for doing this!

File: bin/build_breakpad.sh [hunk 2]
```diff
@@ -15,13 +15,16 @@ set -v -e -x
...
+
+# Locate the local patches
+BREAKPAD_PATCHES="$(pwd)/breakpad-patches"
 
 export MAKEFLAGS
 MAKEFLAGS=-j$(getconf _NPROCESSORS_ONLN)
 
 if [ ! -d "depot_tools" ]; then
-  git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
+  git clone --depth=1 --single-branch https://chromium.googlesource.com/chromium/tools/depot_tools.git
```
Comment (willkg) [inline bin/build_breakpad.sh:27]: Nice! I kept meaning to do this and then kept forgetting.

File: bin/build_breakpad.sh [hunk 3]
```diff
@@ -54,23 +57,30 @@ else
...
 rsync -a --exclude="*.git" ./src "${PREFIX}"/
 
-# NOTE(willkg): Swap these when you're building a pgo profile
-export PGOFLAGS="-fprofile-use=/app/pgo_profile/"
-# export PGOFLAGS="-fprofile-generate=/app/pgo_profile/"
-
 # Configure breakpad for building
-export CXXFLAGS="-g -flto -O3 ${PGOFLAGS}"
-./configure --prefix="${PREFIX}"
+OPTFLAGS="-O3 -flto ${PGOFLAGS} -Wno-coverage-mismatch -Wno-missing-profile"
```
Comment (willkg) [inline bin/build_breakpad.sh:79]: I noticed that `PGOFLAGS` is coming from the caller script which got me to wonder what was set in this script. I think this script should be doing `set -u` so unset variables error out. I can add that later or you could add it in this PR--either is fine.

**Review by willkg — APPROVED** [review]
This looks great! When we merge this, CircleCI will create a docker image and I can test Socorro with that.

I think I should add `set -u` to the top of the script, but I'll do that in a follow-up PR.

Thank you for automating the pgo profile bits!

**Pull request merged**